azure/types.go

```diff
@@ -206,3 +210,71 @@ type VMSSExtensionSpec struct {
 	Version           string
 	ProtectedSettings map[string]string
 }
+
+type (
+	// VMSSVM defines a VM in a virtual machine scale set.
+	VMSSVM struct {
+		ID               string                    `json:"id,omitempty"`
+		InstanceID       string                    `json:"instanceID,omitempty"`
+		Image            infrav1.Image             `json:"image,omitempty"`
+		Name             string                    `json:"name,omitempty"`
+		AvailabilityZone string                    `json:"availabilityZone,omitempty"`
+		State            infrav1.ProvisioningState `json:"vmState,omitempty"`
+	}
+
+	// VMSS defines a virtual machine scale set.
+	VMSS struct {
+		ID        string                    `json:"id,omitempty"`
+		Name      string                    `json:"name,omitempty"`
+		Sku       string                    `json:"sku,omitempty"`
+		Capacity  int64                     `json:"capacity,omitempty"`
+		Zones     []string                  `json:"zones,omitempty"`
+		Image     infrav1.Image             `json:"image,omitempty"`
+		State     infrav1.ProvisioningState `json:"vmState,omitempty"`
+		Identity  infrav1.VMIdentity        `json:"identity,omitempty"`
+		Tags      infrav1.Tags              `json:"tags,omitempty"`
+		Instances []VMSSVM                  `json:"instances,omitempty"`
+	}
+)
+
+// HasModelChanges returns true if the spec fields which will mutate the Azure VMSS model are different.
+func (vmss VMSS) HasModelChanges(other VMSS) bool {
+	equal := cmp.Equal(vmss.Image, other.Image) &&
+		cmp.Equal(vmss.Identity, other.Identity) &&
+		cmp.Equal(vmss.Zones, other.Zones) &&
+		cmp.Equal(vmss.Tags, other.Tags) &&
+		cmp.Equal(vmss.Sku, other.Sku)
+	return !equal
+}
+
+// InstancesByProviderID returns VMSSVMs by ID
+func (vmss VMSS) InstancesByProviderID() map[string]VMSSVM {
+	instancesByProviderID := make(map[string]VMSSVM, len(vmss.Instances))
+	for _, instance := range vmss.Instances {
+		instancesByProviderID[instance.ProviderID()] = instance
+	}
+
+	return instancesByProviderID
+}
+
+// ProviderID returns the K8s provider ID for the VMSS instance
+func (vm VMSSVM) ProviderID() string {
```

Comment:
does this stuff really belong in here?

Reply:
I'm happy to put them where ever. They definitely didn't belong in the `./exp/api` where they were before. 

I do think they belong here as they are basically the spec structs which have yet to be refactored for the exp controllers.